docs/tutorials/typescript/subscriptions/index.md

```diff
@@ -1,416 +1,579 @@
+
+[src/workflows.ts](https://github.com/temporalio/subscription-workflow-project-template-typescript/blob/main/src/workflows.ts)
 
 ```ts
-// /src/workflows.ts
 // ...
-export async function SubscriptionWorkflow(
-  email: string,
-  trialPeriod: string | number
-) {
-  let isCanceled = false;
-  wf.setHandler(cancelSignal, () => void (isCanceled = true)); // new
-  await acts.sendWelcomeEmail(email);
-  if (await wf.condition(() => isCanceled, trialPeriod)) {
-    // reach here if predicate function is true
-    await acts.sendCancellationEmailDuringTrialPeriod(email);
-  } else {
-    // reach here if timeout happens first
-    await acts.sendSubscriptionOverEmail(email);
-  }
-}
+export const cancelSubscription = defineSignal("cancelSubscription");
+// ...
+export const updateBillingChargeAmount = defineSignal<[number]>(
+  "updateBillingChargeAmount"
+);
 ```
 
-Now when you run your `cancel-subscription` script you can see that the cancellation happens immediately.
-And now you know the basics of writing asynchronous time- and Signal-based logic.
-We have documented other patterns you are likely to use, like [sleepUntil](https://docs.temporal.io/typescript/workflows#sleep) and [Updatable Timers](https://docs.temporal.io/typescript/workflows#async-design-patterns).
+<!--SNIPEND-->
+
+This code defines two Signals:
+
+1. `cancelSubscription`: A Signal with no input parameters, used to cancel the subscription.
+2. `updateBillingChargeAmount`: A Signal that takes a single parameter of type number, used to update the billing charge amount.
+
+However, just defining a Signal is just the first step. You also need to let the Workflow know what to do when it receives a Signal of that type. This is done using a Signal Handler, which is a function within the Workflow that listens for incoming Signals of a specified type and defines the Workflow's response.
 
-## Data Model
+In order to implement a Signal handler, you will need to use the [`setHandler`](https://typescript.temporal.io/api/namespaces/workflow/#sethandler) function provided by the TypeScript SDK API. This function associates each Signal with its corresponding handler logic.
 
-We are about to start charging for money and adding more settable properties like `BillingPeriod` and `BillingPeriodChargeAmount`.
-It's time to evolve the data model accordingly.
-We'll define a shared type:
+Within the `workflows.ts` file, implement the Signal handlers as shown below:
+
+<!--SNIPSTART subscription-ts-workflow-definition {"selectedLines": ["45-53"]}-->
+
+[src/workflows.ts](https://github.com/temporalio/subscription-workflow-project-template-typescript/blob/main/src/workflows.ts)
 
 ```ts
-export type Customer = {
-  email: string;
-  trialPeriod: number;
-  billingPeriod: number;
-  maxBillingPeriods: number;
-  initialBillingPeriodCharge: number;
-  id: string;
-};
+// ...
+setHandler(cancelSubscription, () => {
+  subscriptionCancelled = true;
+});
```

Comment:
Since this tutorial is high-level and meant for an audience who may never have  used Temporal before, I don't want to introduce so many new concepts such as cancellations etc. I agree that we should check if subscription is cancelled before billing though! I will instead add

```
      if (subscriptionCancelled) {
        await sendCancellationEmailDuringActiveSubscription(customer);
        return `Subscription finished for: ${customer.id}, Total Charged: ${totalCharged}`;
      }
```
to this block instead of cancellation.